lib/delta-file.js

```diff
@@ -128,3 +140,244 @@ function toTermObjectArray(triples) {
+}
+
+
+
+/**
+ *
+ * @param {*} changeSets deltas
+ * @param {*} config :true if all types should be found for subjects, false if types should only be looked up for subjects without a type
+ * @returns
+ *
+ * None exhaustive type check:
+ * 1. Find all subjects in the statements
+ * 2. Find whether subject has a type within the statements
+ * 3. If not, check other insert/delete statements for the type
+ * 4. If not, query the triplestore to find the type(s) of the subject
+ *
+ * Exhaustive type check:
+ * 1. query the triplestore for the type(s) of the subject
+ */
+function addTypes(changeSets) {
+  const { contextQueries, addTypes } = contextConfig;
+  const { scope, exhaustive } = addTypes;
+
+  let typeCache = new Map();
+
+  if (exhaustive) {
+    for (const changeSet of changeSets) {
+      let { deletes, inserts } = changeSet;
+      inserts = exhaustiveAddTypes(inserts, typeCache);
+      deletes = exhaustiveAddTypes(deletes, typeCache);
+    }
+
+  } else {
+    // find all types in the inserts
+    const subjectTypes = changeSets.map(
+      changeSet => changeSet.inserts
+    ).flat(1).filter(
+      c => c.predicate === RDF_TYPE_URI
+    ).forEach(
+      c => typeCache.set(c.subject, c)
+    );
+
+    for (let changeSet of changeSets) {
+      let { deletes, inserts } = changeSet;
+      // find subjects in the inserts
+      inserts = nonExhaustiveAddTypes(inserts, typeCache);
+      // find subjects in the deletes
+      deletes = nonExhaustiveAddTypes(deletes, typeCache);
+    }
+  }
+  return changeSets;
+}
+
+async function exhaustiveAddTypes(statements, typeCache) {
+  let subjects = [...new Set([...statements.map((statement) => statement.subject)])];
+
+  for (const subject of subjects) {
+    // query the triplestore for the type and add it to the cache
+    typeCache.set(subject, ... await findSubjectTypes(subject));
+    statements.push(typeCache.get(subject));
+  }
+  return statements;
+}
+
+async function nonExhaustiveAddTypes(statements, typeCache) {
+  let subjects = [...new Set([...statements.map((statement) => statement.subject)])];
+
+  for (const subject of subjects) {
+    if (!statements.find(c => c.subject === subject && c.predicate === RDF_TYPE_URI)) {
+      // subject does not have a type
+      // check if the type is in the typeCache
+      if (typeCache.has(subject)) {
+        statements.push(typeCache.get(subject))
+      } else {
+        // query the triplestore for the type and add it to the cache
+        typeCache.set(subject, ... await findSubjectTypes(subject));
+        statements.push(typeCache.get(subject));
+      }
+    }
+  };
+  return statements;
+}
+
+
+
+// QUERYING THE TRIPLESTORE
+const subjectTypeQuery = (graph, subject) => `
+SELECT ?subject ?type
+FROM <${graph}>
+WHERE {
+  VALUES ?subject {
+    ${subject}
+  }
+  ?subject a ?type
+}`
+
+async function findSubjectTypes(subject) {
+  return await querySudo(subjectTypeQuery(INGEST_GRAPH, subject), {}, INGEST_DATABASE_ENDPOINT)
```

Comment:
INGEST_GRAPH is only specific to one type of consumer (https://github.com/lblod/delta-consumer/tree/master/triples-dispatching/single-graph-dispatching) this implies somehow now that the behavior of the consumer will always be linked to one specific graph. which is not necessarily so.

Reply:
Isn't it best to only have one ingest graph per consumer anyway? It's the landing zone where a copy is maintained of the producing source. Dispatching/mapping/combining stuff can be done in the configurable part of the consumer.
I re-used ingest instead of re-inventing another name. However, maybe best to change to `DCR_INGEST` instead?

Reply:
yeah landingzone might sensible. Perhaps best to  document how to use this feature.

Reply:
(and what it assumes)

Reply:
perhaps prefix the newly introduced variables INGEST_GRAPH and NGEST_DATABASE_ENDPOINT with `DCR_` since it is part of the 'core' of the service?

Reply:
@bdevloed @claire-lovisa  thinking a bit further, I don't think that specific variable should be called ingest_graph.
Actually, that graph we specify is the graph to retrieve context? So perhaps rather `DCR_CONTEXT_SOURCE_GRAPH` (better naming welcome)
That would also kinda mitigate the discussion regarding  INGEST_GRAPH limiting potential custom behavior in instances of consumer.
So if, a complex context deduction is needed, it will be limited to one graph for now at the source. My ideas are perhaps not super well explained, but perhaps you get the gist?

Reply:
Oh no i see the maintenance is already done in delta-context.